### ISSUE TITLE
vmselect: data awareness in multi AZ setup

### ISSUE THREAD

**Issue by hagen1778** [VictoriaMetrics/VictoriaMetrics#5496]
### Is your feature request related to a problem? Please describe

vmselect has no data awareness right now. For every received read query it makes a fan-out request to all configured `-storageNode` addresses. It then merges received responses and replies to the user with result.
It is OK to request data from the vmstorage node that doesn't have this data, as their response is supposed to be much faster (index lookup time) compared to vmstorages that do have requested data (index lookup + fetch data from disk + transfer data via network time).

However, some VM installations have vmstorages installed in different geographical regions and each region stores its unique data set. Some regions could be so distant, that the latency of a round-trip request to such regions could exceed the time needed to actually read and transfer data from the closest region. As a result, the average query latency increases because each query will need to wait for this extra latency to the vmstorages in the most distant regions.

Would be great if vmselect could somehow know if some vmstorages could be skipped for each specific query.

### Describe the solution you'd like

One of the approaches I think could be the easiest is to tag vmstorages with some labels and communicate this info to vmselects:
1. Each storage could be instructed with a custom label (`env="prod"`) or list of labels (`env="prod",region="US"` or list of labels with identical keys `env="dev",env="staging"`).
1. When vmselect makes a handshake with configured storage nodes, it gets their labels and stores them in-memory.
1. On query time, vmselect checks filters in the passed query, and if it can match these filters to the storage labels - the query will be sent only to those. This should be applicable to all read API calls, including /series and /label_values.
1. The registry labels: storage  in vmselect's memory will be refreshed on each handshake (re-established connection)

With this approach, user can use a unique `region` label for each of regional VM clusters. Vmselect will be able to extract this labels from vmstorage handshakes. And on query timex if query contains `region` label filter in it vmselect will be able to skip those storages, that don't have this label set.

### Describe alternatives you've considered

Run multiple vmselects, each dedicated to read from a specific region. Use some automation/proxy which will route request from the user to specific vmselect based on regex match against the user's query.

**Comment by hagen1778** [comment]
Related ticket https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7054

**Comment by hagen1778** [comment]
A workaround for this would be using vmauth with [routing by query args](https://docs.victoriametrics.com/vmauth/#routing-by-query-arg). The following vmauth config could help to route query `A: sum(rate(http_requests_total{env="prod"})})` will be routed to `http://cluster-prod`, and query `B: sum(rate(http_requests_total{env="dev"})})`  will be routed to `http://cluster-dev`:
```
unauthorized_user:
  url_map:
    - src_query_args: [ "query=.*env=\"prod\".*" ]
      url_prefix: "http://cluster-prod"
    - src_query_args: [ "query=.*env=\"dev\".*" ]
      url_prefix: "http://cluster-dev"
```
Please note, by default Grafana sends queries via POST method where query args are passed with request body and can't be read by vmauth. In order for vmauth to read GET params, the method should be changed to GET in Grafana datasource settings.

**Comment by valyala** [comment]
I don't like the proposed approach because of the following reasons:

1. It is very easy to mis-configure some `vmstorage` nodes with incorrect set of labels. For example, if some `vmstorage` contains data with labels `env="prod"` and `env="dev"`, while it is configured with `env="prod"` only, then `vmselect` won't send requests with `env="dev"` filter to it, which will result in incomplete query results, which is very hard to debug and fix properly. The problem increases with the number of `vmstorage` nodes with different sets of labels.
2. Such a scheme supposes that the majority of queries at `vmselect` have explicit [label filters](https://docs.victoriametrics.com/keyconcepts/#filtering) on the set of labels configured at the `vmstorage` node. For example, if every `vmstorage` node is configured with `region` label, then it is expected that every query sent to `vmselect` contains `region="..."` filter. If the query doesn't contain the filter on `region` label, or if the query contains some non-trivial filter on the given label such as regexp ( `region=~"complex_regexp_here"`) or negative filter (`region!="foo" or `region!~"complex_regexp_here"`), then `vmselect` needs to query either all the `vmstorage` nodes or a big share of `vmstorage` nodes. This degrades advantages of the feature to the level of the standard behavior of `vmselect` - to send requests to all the `vmstorage` nodes from the `-storageNode` list.
   A classical example of a query without filters on any label is `up == 0` - this is the most common query in alerting rules and in Grafana dashboards, which show unreachable scrape targets. This query doesn't contain any filters on `region`, so it must be sent to all the `vmstorage` nodes.
3. The delay, which is introduced by querying geographically distant `vmstorage` nodes, doesn't exceed a few hundreds of milliseconds (e.g. when querying `vmstorage` nodes in Australia from Europe). This delay is mostly invisible to humans when they load some non-trivial dashboards in Grafana, since users expect that the dashboard need some time to load. This delay doesn't affect alerting and recording rules too much, since it is much lower than the typical interval between the collected samples (aka `scrape_interval`, which usually equals to 10-30 seconds). This delay doesn't increase the load on `vmstorage` nodes, which have no data for the requested [label filters](https://docs.victoriametrics.com/keyconcepts/#filtering). This delay doesn't increase the load on Grafana and vmalert and doesn't multiply with the number of executed queries, since Grafana and vmalert execute queries in parallel.
4. The delay cannot be eliminated when the query contains label filter matching geographically distant `vmstorage` nodes.
5. It is already possible to query different subsets of `vmstorage` nodes in different regions via distinct `vmselect` nodes configured with different subsets of `vmstorage` nodes at `-storageNode` command-line flag. For example, `vmselect1` can be configured with `-storageNode=europe,us`, while `vmselect2` can be configured with `-storageNode=australia,asia`. Then queries for Australia data can be sent to `vmselect2`, while queries for US data can be sent to `vmselect1`.

Given the arguments above, it is better to don't implement data awareness feature in VictoriaMetrics cluster, since it doesn't give visible advantages over the existing functionality, while it introduces a new set of potential issues with the improper configuration of this feature.

A single `vmselect` node is able to efficiently query thousands of `vmstorage` nodes from the `-storageNode` list, so just use this technique. Please file a separate issue if you hit some issues with this approach in production - we'll fix these issues then.

Closing this feature request as won't implement.
